### ISSUE TITLE
System error messages

### ISSUE THREAD

**Issue by Tosin-Balogun** [nhsuk/nhsuk-service-manual-community-backlog#413]
## What
System error message pattern

## Why
We need a consistent pattern to help users recover from system errors. 

These are errors which is more of a system fault but could be triggered by user interaction. For example, error 404 page not found is commonly used pattern for web page system errors

## Anything else
Included the pattern being built at NHS login

# System error developed at NHS login
We developed a series of error screens to help user recover from system related errors
<img width="1237" alt="Screenshot 2021-09-02 at 10 29 22" src="https://user-images.githubusercontent.com/71393402/145414681-5256c9d8-c0ca-4e14-b691-81e2561d4d7e.png">
We tested the middle pattern with 10 users
## Usability testing
We had three hypothesis we wanted to prove or disprove with the card template:
Hypothesis A: If we put a portion of the help centre content on the error screen, would it help reduce users’ confusion about what they need to do?
Hypothesis B: If we added a recovery option like a ‘try again button’ or a ‘refresh link’, would it help users feel confident to recover?
Hypothesis C: If we restyled the help centre link on the page to be more obvious it would take the users outside the page, would it help users know and weigh more options available outside the error screen?

We tested the prototype with 10 people from a mix of different demographics. We found that hypothesis A and B worked as intended.
Users reacted positively to the help being provided there on the error screen. In some cases, they skipped the problem description and went straight to the help card or straight to the recovery ‘try again’ button.
Some users even expected a link to re-centre themselves back to the homepage of the NHS app.

Hypothesis C had ‘mixed’ results. For the most part, it worked in terms of most users recognising what it was, but 2 out of 10 users missed it. We also saw some of those who recognised the link didn’t want to click it because they didn’t trust ‘FAQs’.

When users were encouraged to click to the help centre, they felt confused and disorientated when they got there, which revalidated what the earlier research audit had informed us. We also saw that users showed preference for economic use of words, especially users with reading difficulties.
## Needs further testing
We were able to build guidance on how to apply the pattern based on the usability feedback but more testing is required in order to refine it
<img width="928" alt="Screenshot 2021-07-09 at 11 53 41" src="https://user-images.githubusercontent.com/71393402/145414145-586c7183-75f3-45c4-b2ad-7a32ff9ba7be.png">

You can read more here https://megaspiderweb.medium.com/helping-users-recover-from-an-error-144a0d8b9500

**Comment by Tosin-Balogun** [comment]
Included the equivalent system error pattern on [the GOVUK design system](https://design-system.service.gov.uk/patterns/problem-with-the-service-pages/)

<img width="896" alt="Screenshot 2022-01-17 at 13 25 36" src="https://user-images.githubusercontent.com/71393402/149777114-f7d934ce-0db5-4d96-a11f-d51b942b80b3.png">

**Comment by ZMcClatchey** [comment]
We have error pages across the cervical screening service. We didn't always put the error pages in the prototype we just had that confluence page which our devs used to build the page in the dev environment. 

This is an example from the prototype so the content might not be up to date. I'll see if i can get some examples from the dev environment too

![cervical-screening herokuapp com_errors_defer_prior-notification-failed-to-load](https://user-images.githubusercontent.com/45592606/149786548-8f71ac56-f8e5-4b1d-8547-cdb17416dd02.png)

**Comment by ZMcClatchey** [comment]
This is the error if the Cervical screening management system cant find the nhs number entered. 

![Screenshot 2022-01-17 at 14 35 12](https://user-images.githubusercontent.com/45592606/149789045-a33bea11-7c40-4e9a-a360-28d5394c7d8e.png)

**Comment by nancyhalladay24** [comment]
Error designs for 111 online similarly aim to give the user an alternative way to continue their journey. 

Below is an error where the service the user was trying to refer themselves to has closed, so we recommend they look at other services available:

<img src="https://user-images.githubusercontent.com/63845568/150149589-4e39c34e-4613-49ce-ba06-e1af07e12124.png" alt="alt text" width="200" height="550">

Below is an ITK failure error where we haven't been able to send the users information on to the service. In this instance it is still appropriate for the user to go to the service anyway. But we also suggest they can look at their other recommended services. 

<img src="https://user-images.githubusercontent.com/63845568/150149607-e2acaac0-1822-4d79-a5bb-5f38c9e99661.png" width="200" height="900">

**Comment by nancyhalladay24** [comment]
Also referring to nielson and normans design heuristics is useful here: https://www.nngroup.com/articles/ten-usability-heuristics/

**Comment by sarawilcox** [comment]
We collected some questions, hint text and error messages in 2020. You can view them in NHS.UK's Confluence: https://nhsd-confluence.digital.nhs.uk/pages/viewpage.action?pageId=163247781. 

See, for example, the example error messages for "There is a problem with the service" and "Service unavailable".

**Comment by beccagorton182** [comment]
We redesigned some error messages on NBS when we learned some users thought their 1st dose was booked when they hadn't been able to find a 2nd dose appointment. We made the pages as short as possible, made the H1 an actionable statement and made any actions clearer on the page. Using bullet points was really useful in the content too.
![Dose 2 Option 1](https://user-images.githubusercontent.com/52791357/150335971-0ce790a7-20af-49e1-81f1-f7ec2974b4d4.png)
![None at all option 1](https://user-images.githubusercontent.com/52791357/150335973-3370080f-a8ad-4f38-ab5c-2e09939a5c54.png)
![None at centre - positive action](https://user-images.githubusercontent.com/52791357/150335974-50df0f52-4aaf-4a73-acd5-b70e7947c31b.png)

**Comment by Tosin-Balogun** [comment]
We are currently doing a piece of work to look at error pages on the NHS app

https://github.com/nhsuk/nhsapp-frontend/issues/119

**Comment by anna-rigo** [comment]
# NHS App usability testing

On 13-14th November 2024 we took a HTML prototype to the NHS research lab in Leeds. 

<img width="580" alt="Screenshot" src="https://github.com/user-attachments/assets/c98484fb-574a-4f45-9ee6-c3a0d5f378db">

## Why

> We observed that users do not read information on error pages when a primary button is present
IF we change the primary button to a secondary button
THEN users will read the information on the error pages
We'll know this when spend time interpreting the information 

This hypothesis was disproved

> We observed that GOV guidance advises against using red test to warn people for 'problem with service' error pages. 
IF we remove red warning icon
THEN users will still understand the error 
We'll know this when users correctly interpret the error page 

This hypothesis was proved

## Who we spoke to

* 6 participants
* including 2 participants with red-green colour deficiency and 1 neurodivergent participant
* users and non-user of the NHS App

## Key findings

* 2/6 participants fully read the ‘We could not log you in’ screen unprompted, 4/6 did not fully read it 
* 4/6 were unable to recall any of the two error screens that has a secondary button present
* 6/6 understood what actions to take on the error screen ‘We still couldn’t connect to your GP system’ 
* 3/6 noticed the error code on the ‘We still could not connect to your GP system’, 2/3 said they would try using it to resolve the issue

## Overall findings

* Participants weren’t concerned with understanding why errors occur, or the pages themselves, as they were more concerned about finding a resolution preferably within the NHS App. 
* Participants were more likely to read the information on the error pages above the secondary button, suggesting that the secondary button prompts them to use it as quick resolution / action

## Quotes

* "It doesn't have a refresh" [‘We still couldn’t connect to your GP system’ screen]
* Remember reading anything [first log-in error]? "No, just 'Good afternoon'"
* "I just want to bulldoze my way through" 

## Next steps:

* Explore further research on error pages with and without a button 
* Review our error page guidance in light of the findings
* Design new examples to be used within the guidance

## We'll also

* collaborate with the Service Management team on how we present users with error codes
* explore the possibility of creating some more generic, standard error pages that can be used for a grouping of similar technical problems

**Comment by sarawilcox** [comment]
See the NHS App design system error pages pattern: https://design-system.nhsapp.service.nhs.uk/patterns/error-page/

<img width="543" height="739" alt="Image" src="https://github.com/user-attachments/assets/ecbc3cc3-b35f-4e76-98d5-6517a38295cc" />

**Comment by LauraJD** [comment]
**Community survey findings - problem with this service**

- 72% of respondents have a need for this pattern
- 50% said it would be extremely useful and 50% said it would be very useful to have a common NHS pattern 
- 10 volunteers

**Key features:**
- Provide an alternative way to progress
- Provide a link to navigate elsewhere e.g. to home/start page, site map
- Option to try re-enter details (NHS login)
- How to re-try e.g. refresh the page
- Try again later message
- 'Contact us' option

**Comment by sarawilcox** [comment]
See also timed out pages in [Session time out](https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/479)

**Comment by JessDesignsStuff** [comment]
The Prescription Tracker is a service for NHS clinicians and admin staff to find prescription information that is not in their Patient Medication Record (PMR) system. We have followed GOV.UK error pages for 404, 500 and ‘There is a problem with this service’ scenarios. These pages have followed these principles: 
- explain the error in plain English and be as specific as possible about what has gone wrong, including a short and useful H1 
- give options for the user to fix the issue themselves, or explain how long the issue is expected to last if the problem is with our service 
- include some support or another channel for them to continue their journey, if possible 
Error 404
<img width="1135" height="797" alt="Image" src="https://github.com/user-attachments/assets/af19fb3e-eab0-44d0-b364-f1c63a7f43ec" />
Error 500
<img width="1135" height="797" alt="Image" src="https://github.com/user-attachments/assets/aaf56720-8e36-46f7-a74d-08d0f4984790" />
We have also followed this pattern for unsuccessful searches. Users can search for a patient or prescription and if their search criteria shows no results, the content says that no patients or prescriptions have been found and gives guidance on how to refine their search. 
<img width="1135" height="797" alt="Image" src="https://github.com/user-attachments/assets/2141da71-607e-4ee5-b1c4-78ec1aaa7067" />
